core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,287 @@
+      fileName = file.toUri().toString();
+      fs = file.getFileSystem(conf);
+      XMLInputFactory factory = XMLInputFactory.newInstance();
+      FSDataInputStream fdDataInputStream = fs.open(file);
+      try {
+      reader = factory.createXMLStreamReader(fdDataInputStream);
+      } catch (XMLStreamException exception) {
+        throw new RuntimeException("XMLStreamException exception : ", exception);
+      }
+      //set required node path details.
+      nodePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH);
+      //remove preceding '/' in node path to avoid first unwanted element after split('/')
+      if (nodePath.indexOf("/") == 0) {
+        nodePath = nodePath.substring(1, nodePath.length());
+      }
+      nodes = nodePath.split("/");
+      totalNodes = nodes.length;
+
+      //map to store node path specified nodes with key as node level
+      actualNodeLevelMap = new HashMap<Integer, String>();
+      for (int i = 0; i < totalNodes; i++) {
+        actualNodeLevelMap.put(i, nodes[i]);
+      }
+      currentNodeLevelMap = new HashMap<Integer, String>();
+
+      tempFilePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE);
+      fileAction = conf.get(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION);
+      targetFolder = conf.get(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (reader != null) {
+      try {
+        reader.close();
+      } catch (XMLStreamException exception) {
+        // Swallow exception.
+      }
+    }
+  }
+
+  @Override
+  public float getProgress() throws IOException {
+    return 1.0f;
+  }
+
+  @Override
+  public LongWritable getCurrentKey() throws IOException, InterruptedException {
+    return currentKey;
+  }
+
+  @Override
+  public Map<String, String> getCurrentValue() throws IOException, InterruptedException {
+    return currentValue;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    currentKey = new LongWritable();
+    currentValue = new HashMap<String, String>();
+    int lastNodeIndex = totalNodes - 1;
+    String lastNode = nodes[lastNodeIndex];
+    StringBuilder xmlRecord = new StringBuilder();
+
+    //flag to know if xml record matching to node path has been read and ready to use.
+    boolean xmlRecordReady = false;
+    //flag to know if matching node found as per node path
+    boolean nodeFound = false;
+
+    try {
+      while (reader.hasNext()) {
+        int event = reader.next();
+
+        switch (event) {
+          case XMLStreamConstants.START_ELEMENT:
+            String nodeNameStart = reader.getLocalName();
+            boolean validHierarchy = true;
+            currentNodeLevelMap.put(nodeLevel, nodeNameStart);
+
+            //check if node hierarchy matches with expected one.
```

Comment:
currentNodeLevelMap is getting updated for each node encounter while iterating through reader. It starts with first node and get added to map, then it moves to next node and so on.
Like - 
catalog
	- book
		- price

suppose there are other tags inside book tag like title, author. 
Now loop encounters author tag and it won't match with price then I should exit the loop because we are looking for price tag only(I missed to add break in loop, my bad), also there are less chances to go back to first node again.
This will avoid unnecessary iteration of loop.  
Please let me know your inputs.